File: Sources/Gist/Gist.swift
import DesertAnt

/// A single predicted topic and its probability.
public struct Topic: Sendable, Identifiable, Equatable {
    public var id: String { slug }
    /// The taxonomy slug, e.g. `"technology"`.
    public let slug: String
    /// The human-readable name, e.g. `"Technology & Software"`.
    public let name: String
    /// The model's probability, `0...1`.
    public let score: Double
}

/// Errors thrown while loading or running the model. (`MessageError` is
/// `LocalizedError` wherever Foundation exists, so `localizedDescription`
/// shows `message`.)
public enum GistError: MessageError, Sendable {
    /// A model resource (model, tokenizer, embedding, or metadata) was missing.
    case modelNotFound
    /// On-device prediction failed or returned an unexpected output.
    case predictionFailed

    public var message: String {
        switch self {
        case .modelNotFound: "A Gist model resource was not found."
        case .predictionFailed: "On-device topic tagging failed."
        }
    }
}

/// On-device, multi-label content topic tagging.
///
/// `Gist` tags a post (title, or title + description) with topics from a fixed
/// 36-topic taxonomy across 101 languages, fully on device. A potion embedding
/// lookup and a hashed n-gram stream feed an MLP head that runs through the
/// shared inference session (Core ML on Apple, LiteRT elsewhere). Create one
/// once and reuse it.
///
/// ```swift
/// let gist = Gist()
/// let topics = try await gist.classify("How to start a podcast with your iPhone")
/// // [Topic(slug: "technology", name: "Technology & Software", score: 0.91), ...]
/// ```
public final class Gist: @unchecked Sendable {
    // Resolving the files, loading once, sharing that load, and reporting
    // availability are the same for every model, so they live in the core's
    // `LoadedModel`; Gist adds only how a resolved directory becomes its model.
    private let model: LoadedModel<Model>

    /// Creates a tagger. Construction does no work and starts no download; the
    /// model loads on the first ``classify(_:topK:threshold:)``, ``scores(of:)``
    /// or ``download(progress:)``, off your calling thread.
    ///
    /// `variant` selects the model: ``GistVariant/multilingual`` (the default,
    /// 101 languages) or ``GistVariant/english`` (the ~15 MB English-only build;
    /// English/Latin text only). Each variant downloads and caches only its own
    /// files.
    ///
    /// `directory` is where the model lives. If it already contains the model
    /// (you pre-downloaded or shipped it there) it is used offline; otherwise the
    /// model is downloaded into it and reused offline afterward. With no
    /// `directory` (the default), a managed cache location is used.
    ///
    /// Nothing is bundled with this package. To ship the model with your app,
    /// point `directory` at a folder you populated with the model files: it is
    /// used as-is, offline, and nothing is downloaded.
    public convenience init(variant: GistVariant = .default, directory: String? = nil) {
        self.init(variant: variant, directory: directory, cacheRoot: nil)
    }

    /// Binding entry point that also supplies the platform base cache root under
    /// which the managed layout lives (the app cache dir on Android, node
    /// `~/.cache` on the web). On Apple/Linux FileManager provides it, so the
    /// public `init(variant:directory:)` passes `nil`.
    @_spi(GistBindings)
    public init(variant: GistVariant = .default, directory: String?, cacheRoot: String?) {
        // The variant's own slice of the repo, not the catalog's default
        // manifest, so selecting English never downloads the 74 MB build.
        model = LoadedModel(variant.distribution, directory: directory, cacheRoot: cacheRoot) { files in
            try Model(assets: await .gist(files: files, variant: variant))
        }
    }

    /// Creates a tagger from explicitly provided assets (used by the Android/JNI
    /// and custom-deployment paths).
    @_spi(GistBindings)
    public init(assets: ModelAssets) {
        model = LoadedModel { try Model(assets: assets) }
    }

    /// Whether the model is available for this tagger with no network: cached
    /// (for the managed location) or already present in `directory`.
    public func isDownloaded() -> Bool { model.isDownloaded() }

    /// Download and load the model ahead of time, so the first classify is
    /// instant. Reports download progress `0...1`. Concurrent calls, and an
    /// implicit load from a classify, share one download. A no-op once loaded
    /// (see ``isDownloaded()``).
    public func download(progress: @Sendable @escaping (Double) -> Void = { _ in }) async throws {
        try await model.download(progress: progress)
    }

    /// Await model readiness. The bindings use this to surface load errors
    /// eagerly; apps can just call ``classify(_:topK:threshold:)``.
    @_spi(GistBindings)
    public func waitUntilLoaded() async throws {
        _ = try await model.value()
    }

    /// The full 36-topic probability distribution for `text` (slug -> probability).
    /// Use these for channel roll-up (see ``channelTopics(_:options:)``).
    public func scores(of text: String) async throws -> [String: Double] {
        try await model.value().scores(text)
    }

    /// Everything the cross-language bindings send over the wire in one call:
    /// the distribution plus the tuned threshold and display names that let a
    /// host derive `classify` without shipping `gist_config.json` or
    /// `taxonomy.json`. The public API splits these across two calls and a
    /// loaded `Model`, neither of which crosses the ABI.
    func tagged(_ text: String) async throws -> (threshold: Double, names: [String: String], scores: [String: Double]) {
        let model = try await model.value()
        return (model.threshold, model.names, try await model.scores(text))
    }

    /// The ranked topics for `text` above the model's tuned threshold (the top
    /// topic is always returned). `topK` caps the list (default 3).
    public func classify(_ text: String, topK: Int = 3, threshold: Double? = nil) async throws -> [Topic] {
        let model = try await model.value()
        let thr = threshold ?? model.threshold
        let scores = try await model.scores(text)
        let ranked = scores.sorted { $0.value != $1.value ? $0.value > $1.value : $0.key < $1.key }
        return ranked.prefix(topK).enumerated().compactMap { i, kv in
            (kv.value >= thr || i == 0)
                ? Topic(slug: kv.key, name: model.names[kv.key] ?? kv.key, score: kv.value)
                : nil
        }
    }
}

File: packages/gist-kotlin/src/main/kotlin/ai/desertant/gist/Gist.kt
package ai.desertant.gist

import ai.desertant.core.FfiWriter
import ai.desertant.core.LoadedModel

/** The catalog id, which is how the shared native layer is asked for Gist. */
private const val MODEL_ID = "gist"
private const val MODEL_NAME = "Gist"

/** A single predicted topic and its probability, from [Gist.classify]. */
data class Topic(
    /** The taxonomy slug, e.g. `"technology"`. */
    val slug: String,
    /** The human-readable name, e.g. `"Technology & Software"`. */
    val name: String,
    /** The model's probability, from `0.0` to `1.0`. */
    val score: Double,
)

/** Thrown when the model cannot be created, loaded, or run. */
class GistException(message: String) : Exception(message)

/**
 * On-device, multi-label content topic tagging. Mirrors the iOS/Swift SDK: create
 * one `Gist` and reuse it; the model loads lazily on the first [classify] or
 * [scores] (or eagerly via [download]).
 *
 * ```kotlin
 * val gist = Gist(context)                        // download on demand, cached
 * val topics = gist.classify("How to start a podcast with just your iPhone")
 * // [Topic("technology", "Technology & Software", 0.93), ...]
 * gist.close()
 * ```
 *
 * Creating, downloading, running, and releasing the model are the shared
 * `ai.desertant:core` shell ([LoadedModel]); what lives here is Gist's API and
 * its payload schemas.
 *
 * The English-only model build is currently selectable from the Swift SDK only.
 *
 * @param directory the model's home. Files already there are adopted (so an app
 *   that ships the model just points at the folder it unpacked it into),
 *   otherwise the model is downloaded into it. Omit to use the app cache.
 */
class Gist(
    context: android.content.Context,
    directory: String? = null,
) : AutoCloseable {
    private val model = LoadedModel(MODEL_ID, MODEL_NAME, context, directory, ::GistException, GistNative)

    companion object

    /** Whether the model is available for this tagger with no network. */
    fun isDownloaded(): Boolean = model.isDownloaded()

    /**
     * Download the model ahead of time so the first [classify] is instant. A
     * no-op once available (see [isDownloaded]). Suspends on a background
     * dispatcher.
     */
    suspend fun download() = model.download()

    /**
     * The full 36-topic probability distribution for [text] (slug to
     * probability). Feed these to [channelTopics] to roll a channel up. Empty or
     * blank input returns an empty map.
     */
    suspend fun scores(text: String): Map<String, Double> {
        if (text.isBlank()) return emptyMap()
        return tagged(text).scores
    }

    /**
     * The ranked topics for [text] above the model's tuned threshold, most likely
     * first. The top topic is always returned even when nothing clears the
     * threshold; [topK] caps the list and [threshold] overrides the model's own.
     * Empty or blank input returns an empty list.
     */
    suspend fun classify(text: String, topK: Int = 3, threshold: Double? = null): List<Topic> {
        if (text.isBlank()) return emptyList()
        val tagged = tagged(text)
        val cutoff = threshold ?: tagged.threshold
        // Ranked by score, ties broken by slug - the same order Swift produces.
        return tagged.scores.entries
            .sortedWith(compareByDescending<Map.Entry<String, Double>> { it.value }.thenBy { it.key })
            .take(topK)
            .filterIndexed { i, e -> e.value >= cutoff || i == 0 }
            .map { Topic(it.key, tagged.names[it.key] ?: it.key, it.value) }
    }

    /** One run, decoded. Both public calls need the same payload. */
    private suspend fun tagged(text: String): Tagged {
        // Input payload: the text. Options payload: empty. Result payload: an f64
        // tuned threshold, a count, then per topic a slug, a display name, and an
        // f64 probability - the whole taxonomy, ordered by slug. All three must
        // match Sources/Gist/Binding.swift.
        val input = FfiWriter().string(text).done()
        return model.run(input, failureMessage = "topic tagging failed") { r ->
            val threshold = r.double()
            val count = r.int()
            val scores = LinkedHashMap<String, Double>(count)
            val names = LinkedHashMap<String, String>(count)
            repeat(count) {
                val slug = r.string()
                names[slug] = r.string()
                scores[slug] = r.double()
            }
            Tagged(threshold, scores, names)
        }
    }

    private data class Tagged(
        val threshold: Double,
        val scores: Map<String, Double>,
        val names: Map<String, String>,
    )

    /** Release the native model. The tagger is unusable afterwards; calling this
     *  again is a no-op. */
    @Synchronized override fun close() = model.close()
}
